Mathlib/Combinatorics/Optimization/ValuedCSP.lean

```diff
@@ -20,13 +21,31 @@ General-Valued CSP subsumes Min-Cost-Hom (including 3-SAT for example) and Finit
 * `ValuedCSP.Instance`: An instance of a VCSP problem over given template.
 * `ValuedCSP.Instance.evalSolution`: An evaluation of the VCSP instance for given solution.
 * `ValuedCSP.Instance.IsOptimumSolution`: Is given solution a minimum of the VCSP instance?
+* `Function.HasMaxCutProperty`: Can given binary function express the Max-Cut problem?
+* `FractionalOperation`: Multiset of operations on given domain of the same arity.
+* `FractionalOperation.IsSymmetricFractionalPolymorphismFor`: Is given fractional operation a
+   symmetric fractional polymorphism for given VCSP template?
 
 ## References
 * [D. A. Cohen, M. C. Cooper, P. Creed, P. G. Jeavons, S. Živný,
    *An Algebraic Theory of Complexity for Discrete Optimisation*][cohen2012]
 
 -/
 
+lemma column_of_2x2_left {α : Type*} (a b c d : α) :
```

Comment:
Would you name it `column_of_two_mul_two_left` then? Wouldn't it hint more towards operations on integers (`2*2=4`) then to sizes of matrices?

Reply:
I wouldn't use mul. I might call it `Matrix.fin_two_column_one` or something like that.

But another alternative is just to ditch these lemmas and close the goal with
`convert symmega ![a, b] ![b, a] (List.Perm.swap b a []) o in_omega using 2 <;> ext i <;> fin_cases i <;> rfl`

Reply:
Thanks! I'll rename them.

Reply:
I renamed and moved them. You can check it now.

Reply:
`simp` can almost prove these lemmas by itself. It's only missing 
```
@[simp]
lemma const_fin1_eq {α : Type*} (x : α) : (fun _ : Fin 1 ↦ x) = ![x] := 
  Matrix.cons_fin_one x _ |>.symm
```
Though I'm not sure if that is a good global simp-lemma. 

In any case, you can replace the line `rw [column_of_2x2_left, column_of_2x2_right]` by `simp [← Matrix.cons_fin_one]` and then you don't need these two lemmas.

Reply:
Did you see this comment?

Reply:
Yes, I did. It doesn't seem like improvement.

Reply:
using `simp` seems better than an unbounded number of required lemmas of the form `fin_n_column_m`.

Reply:
OK, I tried your suggestion. However `simp [← Matrix.cons_fin_one]` does not work.

Reply:
What can be done, however, is this:
```
convert symmega ![a, b] ![b, a] (List.Perm.swap b a []) o in_omega using 2 <;>
    simp [Matrix.cons_fin_one]
```

Reply:
I am happy with this version. It works and it doesn't use `simp` nonterminally.

Reply:
You can replace
```
  rw [Matrix.fin_two_column_zero, Matrix.fin_two_column_one]
  exact symmega ![a, b] ![b, a] (List.Perm.swap b a []) o in_omega
```
by
```
  convert symmega ![a, b] ![b, a] (List.Perm.swap b a []) o in_omega using 2 <;>
    ext j <;> fin_cases j <;> rfl
```
but it is a bit obfuscated.

I prefer to have a lemma for such a thing.
What about `private lemma` in my file?

Reply:
`import Mathlib.Data.Matrix.Notation`

Reply:
All right! Let's see whether the whole Mathlib builds now.

Reply:
It doesn't. I'll try to fix it once I return to my office.

Reply:
It depends on how bad the failures are. Feel free to make it a non-simp lemma, and use `simp [const_fin1_eq]` in your case.

Reply:
You know you can get the Mathlib cache also for branches, right? (So you don't have to build locally.)

Reply:
My impression is that it doesn't work on my computer. For example `Mathlib.Analysis.Analytic.Inverse` failed in the CI, which means all of dependencies should be built correctly on the server, right? Yet if I `lake exe cache get` and then open this file in VS Code, its building takes forever.

Reply:
All right, it seems I fixed all the errors caused by adding the new simp lemma. Is there anything else you want me to change?

Reply:
Oh, there is something wrong in the `test` folder now.

Reply:
What command should I use to build this file `https://github.com/leanprover-community/mathlib4/blob/master/test/matrix.lean` locally?

Reply:
I left a TODO in `test/matrix`. I don't want to fix the test if I don't know what the purpose of the test is.

Reply:
This new simp lemmas is breaking things we don't want to break. `simp` should still work on the proofs where it worked before. 
We can do that by adding more simp lemmas (e.g. `![a] + ![b] = ![a + b]` and for many other operations than `+`). Ideally sufficiently many so that all the existing proofs work without modification (i.e. not removing this lemma from the simp-set). This is my preferred solution, but more work.
But feel free to also remove the simp-attribute simp-lemma globally (potentially with a comment `Floris thinks we should make this a global simp lemma, but that requires a bunch of extra simp-lemmas to not break existing simp reductions`).

Reply:
All right, I will add "singleton lemmas" for `+`, `-`, `*`, ˙/˙, what else? Smul?

Reply:
I removed the `@[simp]` tag on `const_fin1_eq`. Let's review this PR as it is. After merging, I'll create another PR that adds the `@[simp]` tag and solves all the related side quests.